operator/cnp_event.go

```diff
@@ -101,3 +130,179 @@ func enableCNPWatcher() error {
+				err := json.Unmarshal(event.Value, &cnpStatusUpdate)
+				if err != nil {
+					log.WithFields(logrus.Fields{"kvstore-event": event.Typ.String(), "key": event.Key}).
+						WithError(err).Error("Not updating CNP Status; error unmarshaling data from key-value store")
+					continue
+				}
+
+				namespace, name, node, err := extractFieldsFromKey(event.Key)
+				if err != nil {
+					log.WithFields(logrus.Fields{"kvstore-event": event.Typ.String(), "key": event.Key}).
+						WithError(err).Error("Not updating CNP Status; error extracting fields from key")
+					continue
+				}
+
+				log.WithFields(logrus.Fields{
+					"name":      name,
+					"namespace": namespace,
+					"node":      node,
+					"key":       event.Key,
+					"type":      event.Typ,
+				}).Debug("received event from kvstore")
+
+				nameNamespace := fmt.Sprintf("%s/%s", namespace, name)
+				mgr.mutex.Lock()
+				ch, ok := mgr.eventMap[nameNamespace]
+				if !ok {
+					ch = make(chan *NodeStatusUpdate, 512)
+					mgr.eventMap[nameNamespace] = ch
+					mgr.mutex.Unlock()
+
+					// Note that the controller below is a closure on this
+					// value. We don't need to worry about synchronizing access
+					// to this map because while its lifecycle is outside of
+					// the controller goroutine, the controller goroutine is
+					// the only goroutine manipulating it.
+					nodeStatusMap := make(map[string]cilium_v2.CiliumNetworkPolicyNodeStatus)
+					controllers.UpdateController(nameNamespace, controller.ControllerParams{
+						DoFunc: func(ctx context.Context) error {
+						Loop:
+							for {
+								select {
+								case <-ctx.Done():
+									// Controller was stopped, we can simply exit.
+									return nil
+								case ev, ok := <-ch:
+									if ok {
+										nodeStatusMap[ev.node] = *ev.CiliumNetworkPolicyNodeStatus
+									}
+								default:
+									break Loop
+								}
+							}
+
+							if len(nodeStatusMap) == 0 {
+								return nil
+							}
+
+							var (
+								cnp *types.SlimCNP
+								err error
+							)
+
+							switch {
+							// Patching doesn't need us to get the CNP from
+							// the store because we can perform patches without
+							// needing the actual CNP object itself.s
+							case k8sversion.Capabilities().Patch:
+							default:
+								cnp, err = getUpdatedCNPFromStore(ciliumStore, nameNamespace)
+								if err != nil {
+									return err
+								}
+							}
+
+							// Now that we have collected all events for
+							// the given CNP, update the status for all nodes
+							// which have sent us updates.
+							if err = k8s.UpdateStatusesByCapabilities(k8s.CiliumClient(), k8sversion.Capabilities(), cnp, namespace, name, nodeStatusMap); err != nil {
+								return err
+							}
+
+							return nil
+						},
+						RunInterval: time.Second * 10,
+						StopFunc: func(ctx context.Context) error {
+							close(ch)
+							mgr.mutex.Lock()
+							delete(mgr.eventMap, nameNamespace)
```

Comment:
I have a feeling that this isn't as simple as removing from the map, as I'm not sure what's stopping a `Create` or `Modify` event from coming in from another node for this same CNP and adding it back into the controller manager after we have stopped this controller in the CNP watcher above...

I was thinking maybe we should wait for a delete event to be received from every single node until there are no more keys left for this CNP in the key-value store (tracked by perhaps a set of all node names that this CNP is aware of), and compare that to the set of nodes in the cluster at the time we receive the delete event from K8s?